Fixes #12716 - Extensible pages for smart proxy with mount points

xprazak2 opened the pull request:
As per today's phonecall with @lzap, I significantly modified the initial page extensibility. I removed the `Tab` and `Column` classes and more generic `Pagelet` took their place. The key elements in this version are mounpoints/extension_points.

`extension_point` - place where page/pagelet can be extended. One page/pagelet can have multiple extension_points

`mountpoint` - place where pagelet attaches itself to a parent, only one per pagelet

At a certain place in a view, child pagelets can be rendered by their mountpoint. This hopefully catches the requirement for high level of freedom when adding a new content to the page. Further description and examples are in services/plugin. There are als o a few minor changes based on comments in previous version, I wanted to keep it for now just to see the difference, hence the new PR. Previous version can be found [here](https://github.com/theforeman/foreman/pull/3023)

Commit:
Fixes #12716 - Extensible pages for smart proxy with mount points

lzap commented:
This looks better. Before I start testing this design, let me summarise what we are trying to do. These are facts about our simple extension framework:
- Pagelet is a named partial.
- Extension point is basically a symbol that holds list of pagelets.
- Page represents a controller/action.
- Extension point can be rendered using a loop as anything (blocks on the page or individual tabs).
- Each pagelet within the list has a priority (number).
- It's basically a flat structure actually (symbol as key - list as value).

Now, when I look at your design, I still see one extra complexity and that is the Page class. I understand that we want to prevent extension points name clash, but since we have controller/action symbol available, we should be able to store pagelets in a flat structure. Do you think that would be possible? Basically then registration would become more simple, without any nesting.

I understand that with Page class there is additional context for extension point symbols, but since the whole controller/action flow has one Rails context, it should be intuitive for developers to name extension points in a way they don't interfere each other.

``` ruby
# this is how I extend tabs
extend_page "smart_proxies/show", :tabbed_menu, :priority => 100, :partial => "smart_proxies/tabs/overview", :name => "Overview"
extend_page "smart_proxies/show", :tabbed_menu, :priority => 200, :partial => "smart_proxies/tabs/services", :name => "Services"
# and this is how to extend tab content (or any other content)
extend_page "smart_proxies/show", :foreman_details, :partial => "myplugin/content", :name => "XYZ"
```

We can create some syntactic sugar indeed, maybe `extend_page` can be a DSL block, but no extra nesting, just to make it shorter. One block = one extension point symbol context, this will be pretty intuitive.

What you think?

I am going to provide few more comments for the code itself.

lzap commented on app/controllers/concerns/foreman/controller/extensible_page.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  module ClassMethods
+    def register_extensible_pages(pages)
```
This can accept an array as kargs.

lzap commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -1,5 +1,8 @@
 class SmartProxiesController < ApplicationController
   include Foreman::Controller::AutoCompleteSearch
+  include Foreman::Controller::ExtensiblePage
```
I'd rather use `PageletSupport` so we get rid of Page term completely which is quite misleading.

lzap commented on app/services/pages/loader.rb:
```diff
@@ -0,0 +1,23 @@
+      end
+    end
+  end
+end
```
I think we can completely get rid of the pre-initialization.

Let's suppose that if you try to render an extension point that has no pagelets, you simply return an empty array. Therefore nothing will be rendered, as simple as that. Remove also the early errors when things are not present.

The only benefit I see in having this is to be able to generate list of extensions, but is it worth the effort?

lzap replied:
Also please rename this directory to `pagelets`.

lzap commented on app/services/pages/page.rb:
```diff
@@ -0,0 +1,15 @@
+      "#{@url_hash[:controller]}/#{@url_hash[:action]}".to_sym
+    end
+  end
+end
```
This class is no longer necessary and you can also get rid of inheritance completely.

lzap commented on app/services/pages/pagelet.rb:
```diff
@@ -0,0 +1,22 @@
+      @priority = priority
+      @parent = parent
+      super()
+    end
```
Parent can go as well.

lzap commented on app/controllers/concerns/foreman/controller/extensible_page.rb:
```diff
@@ -0,0 +1,19 @@
+  def find_page
+    @page = ::Pages::Manager.find_page(controller_name, action_name)
+    raise Foreman::Exception.new(N_("Page %s has no view assigned", @page.name)) unless @page || @page.view
+  end
```
This can be renamed as `find_pagelet_context`.

lzap commented:
For the record, I started a feature wiki page here: http://projects.theforeman.org/projects/foreman/wiki/Pagelets

lzap commented on app/controllers/smart_proxies_controller.rb:
```diff
@@ -1,5 +1,8 @@
   include Foreman::Controller::AutoCompleteSearch
+  # include Foreman::Controller::ExtensiblePage
+
+  # register_extensible_pages [:show]
```
Commented intentionally?

lzap commented:
Would you mind rebasing on top of develop? PatternFly was merged already :-(

lzap commented on app/services/pagelets/manager.rb:
```diff
@@ -0,0 +1,35 @@
+        raise ::Foreman::Exception.new("Cannot add pagelet to page #{page_name} without partial") unless opts[:partial]
+        raise ::Foreman::Exception.new("Cannot add pagelet to page #{page_name} without mountpoint") if mountpoint.nil?
+        priority = opts[:priority] || default_pagelet_priority(page_name, mountpoint) - 100
+        siblings = @pagelets[page_name][mountpoint] << Pagelets::Pagelet.new(opts[:name], opts[:partial], priority)
```
Variable `siblings` unused?

lzap commented on app/services/pagelets/manager.rb:
```diff
@@ -0,0 +1,35 @@
+        handle_empty_keys_for page_name, mountpoint
+        raise ::Foreman::Exception.new("Cannot add pagelet to page #{page_name} without partial") unless opts[:partial]
+        raise ::Foreman::Exception.new("Cannot add pagelet to page #{page_name} without mountpoint") if mountpoint.nil?
+        priority = opts[:priority] || default_pagelet_priority(page_name, mountpoint) - 100
```
Why decreasing? I thought that priority is in ascending order, therefore I'd expect `+ 100`. Is it in descending?

xprazak2 replied:
It is descending, should be ascending?

lzap replied:
I guess so, this feels more natural to me: 1, 2 3 :-)

One more comment for this block - can you also store whole `options` hash in Pagelet itself? I plan to pass some variables to my plugin from the initializer block and I want to use it on the ERB page.

lzap commented on app/services/pagelets/manager.rb:
```diff
@@ -0,0 +1,35 @@
+
+      def default_pagelet_priority(page_name, mountpoint)
+      # We need a default priority value for the first pagelet if it is not specified
+        @pagelets[page_name][mountpoint].map(&:priority).min || 2100
```
See above, this is weird, I'd expect zero here increasing by 100.

xprazak2 commented:
I deleted bunch of files, got rid of initialization, pagelets are now within manager in a hash, rebased on patternfly Will do a cleanup and fix tests.

lzap commented on app/services/pagelets/pagelet.rb:
```diff
@@ -0,0 +1,19 @@
+    end
+
+    def <=>(other)
+      other.priority <=> priority
```
I guess you just need to swap these guys in order to get the ascending (more logical) order.
- 0 - first element
- 100 - second
- etc

lzap commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -7,6 +7,9 @@
     <% if service_features.any? %>
       <li><a href="#services" data-toggle="tab"><%= _('Services') %></a></li>
     <% end %>
+    <% sorted_pagelets_at(:tabbed_menu).each do |tab| %>
```
Nitpick: This helper could be named like:

```
<% extension_point_for(:tabbed_menu).each do |pagelet| %>
```

I also kinda don't like `tabbed_menu` name, how about:

```
<% extension_point_for(:main_tabs).each do |pagelet| %>
```

I suggest to use plurals for all extension point names (as they are all collections).

lzap commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -52,6 +55,9 @@
       <div class="col-md-6">
+      <% sorted_pagelets_at(:katello_proxy_details).each do |pagelet| %>
+        <%= render pagelet.partial, {:smart_proxy => @smart_proxy } %>
+      <% end %>
```
We should avoid naming plugins in extension point names, discovery can add something here as well. I suggest `details_contents` or similar.

lzap commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -69,5 +75,10 @@
     <% end %>
+    <% sorted_pagelets_at(:tabbed_menu).each do |tab| %>
+      <div id="<%= tab.snake_name %>" class="tab-pane">
+        <%= render tab.partial, { :smart_proxy => @smart_proxy, :pagelet => tab }%>
```
Yeah, I like you pass pagelets in. Can you do the same for all extension points?

Best approach would be to create `render_pagelet` helper that would do this for us accepting all variables that must be sent in.

lzap commented on app/services/foreman/plugin.rb:
```diff
@@ -175,6 +175,14 @@ def delete_menu_item(menu, item)
+    # extend_page "smart_proxies/show", :foreman_details, :name => "Example Content", :partial => "path/to/partial"
+    def extend_page(page_name, mountpoint, opts)
+      Pagelets::Manager.add_pagelet(page_name, mountpoint, opts)
+    end
```
Nit pick - you could do `extend_page` DSL block for a nicer code, but not necessary really. This is fine.

lzap commented:
@xprazak2 looks great, I have no other comments, going to test this now. I guess you can start working on unit tests as we are under time pressure.

lzap commented:
The patch works as expected.

xprazak2 commented:
I changed the priorities into asc, fixed tests, addded helpers + some minor style improvements

lzap commented on app/services/pagelets/manager.rb:
```diff
@@ -0,0 +1,37 @@
+      def add_pagelet(page_name, mountpoint, opts)
+        handle_empty_keys_for page_name, mountpoint
+        raise ::Foreman::Exception.new("Cannot add pagelet to page #{page_name} without partial") unless opts[:partial]
+        raise ::Foreman::Exception.new("Cannot add pagelet to page #{page_name} without mountpoint") if mountpoint.nil?
```
Please follow http://projects.theforeman.org/projects/foreman/wiki/Translating format.

lzap commented on app/services/pagelets/manager.rb:
```diff
@@ -0,0 +1,37 @@
+
+      def default_pagelet_priority(page_name, mountpoint)
+      # We need a default priority value for the first pagelet if it is not specified
+        priority = @pagelets[page_name][mountpoint].map(&:priority).max ? priority + 100 : 0
```
Out of indentation (comment)

lzap commented on app/views/smart_proxies/show.html.erb:
```diff
@@ -52,6 +53,7 @@
       <div class="col-md-6">
+      <%= render_pagelets_at(:details_content, :smart_proxy => @smart_proxy) %>
         <!--place holder for Capsule sync-->
       </div>
```
Get rid of the comment there, also fix indentation.

lzap commented on test/test_helper.rb:
```diff
@@ -95,9 +95,9 @@ def clear_current_user
-    end
+    # class <<self
+    #   alias_method :test,  :it
+    # end
```
Leftover?

lzap commented on app/services/pagelets/manager.rb:
```diff
@@ -0,0 +1,37 @@
+        raise ::Foreman::Exception.new("Cannot add pagelet to page #{page_name} without partial") unless opts[:partial]
+        raise ::Foreman::Exception.new("Cannot add pagelet to page #{page_name} without mountpoint") if mountpoint.nil?
+        priority = opts[:priority] || default_pagelet_priority(page_name, mountpoint)
+        @pagelets[page_name][mountpoint] << Pagelets::Pagelet.new(opts[:name], opts[:partial], priority)
```
And can you pass `opts` hash and store it in Pagelets as well?

lzap commented on app/helpers/pagelets_helper.rb:
```diff
@@ -0,0 +1,30 @@
+    result = ""
+    sorted_pagelets_at(mountpoint).each do |pagelet|
+      result += "<div id='#{pagelet.snake_name}' class='tab-pane'>"
+      result += (render pagelet.partial, opts.merge!({ :pagelet => pagelet }))
```
Just a nitpick - this is unusual syntax, rubocop might fail.

lzap commented on app/services/pagelets/pagelet.rb:
```diff
@@ -0,0 +1,20 @@
+      @name = name
+      @partial = partial
+      @priority = priority
+      @opts = opts
```
You could delete name and partial from the hash, but not big deal.

lzap commented:
Jenkins builder seems to be down... let's wait.

lzap commented on app/helpers/pagelets_helper.rb:
```diff
@@ -0,0 +1,30 @@
+    end
+    result.html_safe
+  end
+end
```
Can you add one more method that simplifies the actual render line? So rendering in my testing case will be as easy as:

``` ruby
  <% sorted_pagelets_at(:probe_contents).each do |pagelet| %>
    <div class="col-md-6">
      <%= render_pagelet(pagelet, :smart_proxy => @smart_proxy) %>
    </div>
  <% end %>
```

xprazak2 commented on app/services/pagelets/manager.rb:
```diff
@@ -0,0 +1,37 @@
+        raise ::Foreman::Exception.new(N_("Cannot add pagelet to page %s without partial", page_name)) unless opts[:partial]
+        raise ::Foreman::Exception.new(N_("Cannot add pagelet to page %s without mountpoint", page_name)) if mountpoint.nil?
+        priority = opts[:priority] || default_pagelet_priority(page_name, mountpoint)
+        @pagelets[page_name][mountpoint] << Pagelets::Pagelet.new(opts.delete(:name), opts.delete(:partial), priority, opts)
```
@lzap, removing name and partial from opts is already solved

lzap commented on app/services/pagelets/pagelet.rb:
```diff
@@ -0,0 +1,20 @@
+      @name.gsub(/\s+/, "_").underscore
+    end
+  end
+end
```
Mind adding `method_missing` returning an option from the `opt` hash? So instead of

```
pagelet.opts[:probe_name]
```

I can write

```
pagelet.probe_name
```

lzap commented on app/services/pagelets/manager.rb:
```diff
@@ -0,0 +1,47 @@
+      def default_pagelet_priority(page_name, mountpoint)
+        # We need a default priority value for the first pagelet if it is not specified
+        priority = @pagelets[page_name][mountpoint].map(&:priority).max
+        priority.nil? ? 0 : priority + 100
```
This could be more ellegant in a form of:

```
@pagelets[page_name][mountpoint].map(&:priority).push(0).max
```

lzap replied:
Please add a test case when priority is not provided.

lzap commented:
Ok it looks great, plugins can:
- add own pagelets
- add new extensions
- dynamically add pagelets (according to configuration)

I have tested all three cases with my little plugin:

![image](https://cloud.githubusercontent.com/assets/49752/12195619/d0ae49ba-b5f8-11e5-87fe-8c529b60dad1.png)

lzap commented on app/services/pagelets/manager.rb:
```diff
@@ -0,0 +1,47 @@
+        @pagelets[page_name][mountpoint] ||= []
+      end
+
+    end
```
Extra empty line.

lzap commented:
Please to the final rebase and we are good to go. I'll update the docs in the meantime: http://projects.theforeman.org/projects/foreman/wiki/Pagelets

domcleal commented:
The commit message must "fix" an unshipped ticket, never refs.  There is currently a second PR at #3023 which uses the ticket, they must be different.

domcleal commented on app/services/foreman/plugin.rb:
```diff
@@ -175,6 +175,16 @@ def delete_menu_item(menu, item)
+    # extend_page("smart_proxies/show") do |context|
+    #   context.add_pagelet :mountpoint, :name => "Example Pagelet", :partial => "path/to/partial", :priority => 10000
+    # end
+    def extend_page(page_name, &block)
```
Missing test coverage.  It must also be documented on the plugins wiki page.

tstrachota commented:
I'm missing a possibility to display pagelets conditionally. Currently they can only be registered at the plugin load time and there's no option to hide them according to current context (eg. listing environments makes sense only for proxies with pulp node feature). It's problematic only for tab pagelets because they always render headings. Otherwise it works for me in terms of functionality.

IMHO it's not a blocker for this PR so I'll create a separate issue.

tstrachota commented:
Here it is http://projects.theforeman.org/issues/13079

lzap commented:
@xprazak2 fyi tests failed, also fix the commit message as @domcleal noticed.

I will do the documentation and make sure it is referenced from the plugins wiki page.

lzap commented on app/helpers/pagelets_helper.rb:
```diff
@@ -0,0 +1,35 @@
+  def render_pagelets_for(mountpoint, opts = {})
+    result = ""
+    pagelets_for(mountpoint).each do |pagelet|
+      result += (render pagelet.partial, opts.merge!({ :pagelet => pagelet }))
```
This could use `render_pagelet`?

lzap commented:
FYI tests failed. I noticed that there is no mount point for the Overview tab. Maybe you want to add one?

lzap commented:
@shlomizadok there are just two nitpicks and unit tests, feel free to base the features on this branch. No big changes expected.

tstrachota commented on app/helpers/pagelets_helper.rb:
```diff
@@ -0,0 +1,35 @@
+  def render_tab_content_for(mountpoint, opts = {})
+    result = ""
+    pagelets_for(mountpoint).each do |pagelet|
+      result += "<div id='#{pagelet.snake_name}' class='tab-pane'>"
```
I discovered one problem. The `snake_name` should be provided only as a default id with a possibility to override. Currently, javascript fails to switch to pagelet tabs that have name conflicting with an id of a different element in the page structure. For example it's impossible to show pagelet named "Content", because it conflicts with the main '#content' div.

lzap replied:
Perhaps some kind of prefix could make it more safe, we just need an unique name. So `pagelet-name` could be unique enough.

xprazak2 commented:
@tstrachota, I added new `id` option that will be used as a tab id if specified

lzap commented:
Updated documentation at http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Adding-a-Pagelet
and filed template PR at https://github.com/theforeman/foreman/pull/3025

xprazak2 commented:
[test]

domcleal commented:
The test failure is unrelated, please don't retest for known issues as Jenkins load is high enough as it is.

<pre>
11:56:12  12) Failure:
11:56:12 HostIntegrationTest::edit page#test_0004_correctly override global params [/var/lib/workspace/workspace/test_develop_pr_core@2/database/postgresql/label/fast/ruby/1.9.3/test/integration/host_test.rb:452]:
11:56:12 Failed refutation, no message given
</pre>

(http://projects.theforeman.org/issues/13111)

lzap commented:
@xprazak2 deleted your "test" comment to save some resources, please add the `pagelet-` id prefix for default HTML id and I will test and merge running tests locally.

domcleal commented:
If the PR has changed, please continue to use Jenkins and let it run.  Just please don't re-test for a single matrix failure for one documented known issue (at least when there's a queue like today).

lzap commented on app/services/pagelets/pagelet.rb:
```diff
@@ -0,0 +1,24 @@
+    end
+
+    def snake_name
+      opts[:id] || @name.gsub(/\s+/, "_").underscore
```
Wait, I don't like this at all. We are mixing two things together, this should be generic and clear enough to be used in arbitrary snippets.

So I suggest to rename this method to `pagelet_id` or simply `id` and also prefix the default id:

``` ruby
def id
  opts[:id] || 'pagelet-id-' + @name.gsub(/\s+/, "_").underscore
end
```

tstrachota replied:
@lzap :+1: IMHO better

lzap commented on app/services/foreman/plugin.rb:
```diff
@@ -175,6 +175,16 @@ def delete_menu_item(menu, item)
+    # Usage:
+    #
+    # extend_page("smart_proxies/show") do |context|
+    #   context.add_pagelet :mountpoint, :name => "Example Pagelet", :partial => "path/to/partial", :priority => 10000, :id => 'tab_id'
```
Also update the example with `:id => "custom-html-id"` which is more clear.

ShimShtein commented:
How are you dealing with _form extensions? You have to pass the form variable `f` to the pagelets if they want to add fields to the existing form.

ShimShtein commented:
In host unification I have a [similar mechanism](https://github.com/ShimShtein/foreman/blob/puppet-facet/config/initializers/puppet_facet.rb#L5), IMHO it would be better to define the configuration in a single place (facet extensions to the host model and to the view). In this case, can I utilize this framework and will it be a good idea?

xprazak2 commented:
``` Ruby
<%= render_pagelet(pagelet, :f => f) %>
```

should do the trick I think, have a look [here](https://github.com/theforeman/foreman/pull/3025/files#diff-2e05ad7f6b6cd8ca1e79937af504ec78R32)

lzap commented:
Ok running tests locally and then we can merge. Id are fixed, everything is working as expected.

Just to clear: Proxy visibility features do _not_ depend on this patch, features are expected to use ERB normally. This only blocks plugins that want to extend smart proxy page. Although Foreman Core itself can use Pagelets, it is not the best way to do this. @ohadlevy @tstrachota @tbrisker

lzap commented:
Except the failing one all passed:

```
HostgroupIntegrationTest::JS enabled tests#test_0002_parameters change after parent update:
Capybara::ElementNotFound: Unable to find link "Hostgroup"
    test/integration/hostgroup_test.rb:101:in `block (2 levels) in <class:HostgroupIntegrationTest>'
```

lzap commented:
Merged as 33f7c66c311da45e242fc0034d5cf2ebd7a7ffec, thank you!

xprazak2 commented:
@ShimShtein: not sure I follow entirely, but by rendering pagelets for a certain mountpoint in your view, you allow plugins to add content to the page (pagelets attached to a mountpoint get rendered), there is nothing more to it. We have no registration so you would need additional layer that takes care of that. I think it certainly is possible but I cannot see that far to say if it would work well enough.

ShimShtein commented:
1. IMHO this feature will create a difference between API and UI (you will see more properties in UI than in the API). It's worth looking into my implementation in [host unification](https://github.com/ShimShtein/foreman/blob/puppet-facet/app/views/api/v2/hosts/show.json.rabl#L21), maybe you can copy this mechanism, and solve the API-UI conflict.
2. It looks to me pretty similar to `#content_for`, can you tell me what is the difference between your mechanism, and the generic one?

xprazak2 commented:
1. The main focus here was on UI, so I will have to think about how to approach the difference when it happens. I will have a look host unification more closely, thx for the link.
2. #content_for stores a block of markup in an identifier for later use, pagelets can store multiple blocks of markup independently for an identifier, they are also scoped by page
